torch/fx/passes/fuser_utils.py

```diff
@@ -0,0 +1,228 @@
+def copy_module_attributes(gm: torch.fx.GraphModule, subgraph: torch.fx.Graph) -> HolderModule:
+    # Loop through all module calls (call_module) and param fetches (get_attr)
+    # in this component, creating HolderModules as necessary to match the path.
+    # e.g. if in the original module there's a get_attr node fetches "conv.weight".
+    # We create a HolderModule as root -> add a HolderModule named "conv" ->
+    # make "weight" a attribute of "conv" HolderModule and point to conv.weight in
+    # the original module.
+    submodule = HolderModule({})
+    for n in subgraph.nodes:
+        if n.op not in ("call_module", "get_attr"):
+            continue
+
+        target = n.target
+        assert isinstance(target, str)
+        target_name_parts = target.split(".")
+        curr = submodule
+        orig_gm = gm
+
+        for name in target_name_parts[:-1]:
+            if not hasattr(curr, name):
+                curr.add_module(name, HolderModule({}))
+
+            curr = getattr(curr, name)
+            orig_gm = getattr(orig_gm, name)
+
+        leaf_node_name = target_name_parts[-1]
+        leaf_node = getattr(orig_gm, leaf_node_name)
+
+        # Relies on custom __setattr__ magic.
+        setattr(curr, leaf_node_name, leaf_node)
+
+    return submodule
+
+def fuse_partition(gm: torch.fx.GraphModule,
+                   nodes: NodeList,
+                   partition_name: str) -> torch.fx.GraphModule:
+    # returns a graph module that is a copy of `nodes` in gm
+    # assumption: nodes are already sorted in topo order
+
+    for node in nodes:
+        assert node.graph.owning_module is gm, f"{node} doesn't belong to passed in graph module {gm._get_name()}"
+        assert not node._erased, f"{node} has been removed from owning graph"
+        assert node in gm.graph.nodes, f"{node} is not found in graph module {gm._get_name()}"
+
+    # validates partition doesn't introduce dependency circles in the graph
+    assert validate_partition(nodes), "Invalid partition, found dependency cycles"
+
+    subgraph = torch.fx.Graph()
+
+    node_to_placeholder = {}  # mapping of nodes from old graph to placeholder in new graph
+    node_map = {}       # mapping of nodes from old graph to new graph
+
+    # handles inputs throught graph.node_copy's arg_transform functions
+    def remap_inputs(x):
+        if x.op == "get_attr":
+            # TODO: do we really need copy the get_attr node into the graph?
+            # do something here
+            pass
+
+        if x in nodes:
+            # x is inside subgraph, return the copied node
+            # the node should have been copied aleady, as we are copying graph in the topological order
+            return node_map[x]
+
+        if x not in node_to_placeholder:
+            # x is not in subgraph, create a new placeholder for subgraph
+            node_to_placeholder[x] = subgraph.placeholder(x.name, type_expr=x.type)
+
```

Comment:
if we want the placeholder of the subgraph to have tensor meta, can we do this?

`node_to_placeholder[x].meta["tensor_meta"] = x.meta["tensor_meta"]`

Reply:
Yes, we should probably group these into a copy_node_as_placeholder util function, just like FX.Graph.node_copy() function.

Reply:
addressed in latest iteration.